client/src/main/java/org/eclipse/hono/client/impl/DeviceSpecificCommandConsumer.java

```diff
@@ -14,116 +14,90 @@
-     * @param localCloseHandler A handler to be invoked after the link has been closed
-     *                     at this peer's request using the {@link #close(Handler)} method.
-     *                     The handler will be invoked with the link's source address <em>after</em>
-     *                     the link has been closed but <em>before</em> the handler that has been
-     *                     passed into the <em>close</em> method is invoked.
-     * @param remoteCloseHandler A handler to be invoked after the link has been closed
-     *                     at the remote peer's request. The handler will be invoked with the
-     *                     link's source address.
-     * @return A future indicating the outcome of the creation attempt.
-     * @throws NullPointerException if any of the parameters are {@code null}.
+     * @param delegateSupplier Supplies the command consumer that wraps the actual Proton receiver.
+     * @param deviceId The device identifier.
+     * @throws NullPointerException If delegateSupplier or deviceId is {@code null}.
      */
-    public static final Future<DeviceSpecificCommandConsumer> create(
-            final HonoConnection con,
-            final String tenantId,
-            final String deviceId,
-            final Handler<CommandContext> commandHandler,
-            final Handler<String> localCloseHandler,
-            final Handler<String> remoteCloseHandler) {
-
-        Objects.requireNonNull(con);
-        Objects.requireNonNull(tenantId);
-        Objects.requireNonNull(deviceId);
-        Objects.requireNonNull(commandHandler);
-        Objects.requireNonNull(remoteCloseHandler);
-
-        LOG.trace("creating new command consumer [tenant-id: {}, device-id: {}]", tenantId, deviceId);
+    public DeviceSpecificCommandConsumer(
+            final Supplier<DestinationCommandConsumer> delegateSupplier, final String deviceId) {
+        this.delegateSupplier = Objects.requireNonNull(delegateSupplier);
+        this.deviceId = Objects.requireNonNull(deviceId);
+    }
 
-        final String address = ResourceIdentifier.from(CommandConstants.NORTHBOUND_COMMAND_LEGACY_ENDPOINT, tenantId, deviceId).toString();
+    private DestinationCommandConsumer getDelegate() {
+        return delegateSupplier.get();
+    }
 
-        final AtomicReference<ProtonReceiver> receiverRef = new AtomicReference<>();
+    /**
+     * {@inheritDoc}
+     * <p>
+     * Calls the appropriate close handler and closes the outer consumer instance if that contains
+     * no further handlers.
+     */
+    @Override
+    public void close(final Handler<AsyncResult<Void>> closeHandler) {
+        final DestinationCommandConsumer delegate = getDelegate();
+        if (delegate != null) {
+            delegate.removeHandlerAndCloseConsumerIfEmpty(deviceId, closeHandler);
+        } else {
+            LOG.debug("cannot delegate close() invocation; actual consumer not available [consumer device-id {}]",
```

Comment:
Can this actually happen? If so, shouldn't the closeHandler be completed with a succeeded future in this case? Otherwise the calling client would potentially wait forever for the (no longer existing) consumer to be closed, wouldn't it?

Reply:
Before the fix in my second commit (https://github.com/eclipse/hono/pull/1552/commits/303db725b531dab45f28f8ee9af4bbbd29a5013b), this did actually happen, yes.
And you are right, not completing the closeHandler here, has probably caused the delays that caused the test timeouts. 
I've fixed that now, failing the closeHandler.

Reply:
@calohmn WDYT?

Reply:
It still can return `null`. I've added this javadoc for `getDelegate`:
````
    /**
     * Gets the current consumer delegate.
     * <p>
     * {@code null} may be returned here if {@link #close(Handler)} has already been invoked on this
     * consumer (and the delegate consumer got closed as a consequence) or if the delegate consumer
     * was remotely closed and no re-creation of the consumer (e.g. as part of a liveness check) has
     * happened (yet).
     *
     * @return The consumer instance or {@code null}.
     */
````